File: Android/shiba/src/main/java/moe/tlaster/shiba/visitors/ValueVisitor.kt
package moe.tlaster.shiba.visitors

import android.view.ViewGroup
import com.fasterxml.jackson.databind.node.ArrayNode
import com.fasterxml.jackson.databind.node.ObjectNode
import com.fasterxml.jackson.databind.node.ValueNode
import moe.tlaster.shiba.*
import moe.tlaster.shiba.common.Singleton
import moe.tlaster.shiba.converters.RawConverter
import moe.tlaster.shiba.converters.ShibaConverterParameter
import moe.tlaster.shiba.converters.SingleBindingFunctionConverter
import moe.tlaster.shiba.dataBinding.ShibaBinding
import moe.tlaster.shiba.extensionExecutor.IMutableExtensionExecutor
import moe.tlaster.shiba.mapper.ComponentMapper
import moe.tlaster.shiba.mapper.IAllowChildViewMapper
import moe.tlaster.shiba.type.Property
import moe.tlaster.shiba.type.ShibaExtension
import moe.tlaster.shiba.type.ShibaFunction
import moe.tlaster.shiba.type.ShibaObject
import org.liquidplayer.javascript.*


internal object ValueVisitor {
    fun visit(item: Any?, context: IShibaContext?): Any? {
        return when (item) {
            is ShibaView -> visit(item, context)
            is ShibaExtension -> visit(item, context)
            is ShibaFunction -> visit(item, context)
            is ObjectNode -> visit(item, context)
            is ArrayNode -> visit(item, context)
            is ValueNode -> visit(item, context)
            is JSValue -> visit(item, context)
            null -> null
            else -> item
        }
    }

    private val shibaHostMapper by lazy {
        ComponentMapper()
    }

    private fun visit(tree: ShibaView, context: IShibaContext?): NativeView {
        val mapper = Shiba.viewMapping.filter { tree.viewName == it.key }.values.firstOrNull()

        if (context == null) {
            throw IllegalArgumentException()
        }
        if (mapper == null) {
            if (Shiba.components.contains(tree.viewName)) {
                tree.properties.add(Property("componentName", tree.viewName))
                return shibaHostMapper.map(tree, context)
            }
            throw ClassNotFoundException("${tree.viewName} not found")
        } else {
            val target = mapper.map(tree, context)
            if (tree.children.any() && target is ViewGroup) {
                val commonProps = ArrayList<Triple<ShibaView, NativeView, List<Property>>>()
                tree.children.forEach {
                    val child = visit(it, context)
                    if (mapper is IAllowChildViewMapper) {
                        mapper.addChild(target, child)
                    } else {
                        target.addView(child)
                    }
                    val comprop = it.properties.filter { Shiba.configuration.commonProperties.any { cp -> cp.name == it.name } }.toList()
                    if (comprop.any()) {
                        commonProps.add(Triple(it, child, comprop))
                    }
                }
                commonProps.forEach {
                    it.third.forEach { prop ->
                        Shiba.configuration.commonProperties.filter { cp -> cp.name == prop.name }.forEach { cp -> cp.handle(prop.value, it.second, target) }
                    }
                }
            }
            return target
        }

    }

    private fun visit(tree: ShibaExtension, context: IShibaContext?): Any? {
        val executor = Shiba.configuration.extensionExecutors.firstOrNull { it.name == tree.type }
        if (executor != null) {
            val value = executor.provideValue(context, tree)
            if (tree.scriptName.isNullOrEmpty() || tree.scriptName.isNullOrBlank()) {
                return value
            }
            val func = ShibaFunction(tree.scriptName).apply {
                parameter = arrayListOf(tree)
            }
            return when (value) {
                is ShibaBinding -> value.apply {
                    parameter = if (converter != null) {
                        ShibaConverterParameter(converter, parameter, func)
                    } else {
                        func
                    }
                    converter = Singleton.get<SingleBindingFunctionConverter>()
                }
                else ->
                    ShibaBinding("").apply {
                        converter = Singleton.get<RawConverter>()
                        parameter = Singleton.get<SingleBindingShibaFunctionExecutor>().execute(func, tree)
                    }
            }
        }
        throw NotImplementedError()
    }

    private fun visit(tree: ShibaFunction, context: IShibaContext?) : ShibaBinding {
        val function = parseFunction(tree, context)
        val extensions = getExtensions(function)
        if (!extensions.any()) {
            return ShibaBinding("").apply {
                converter = Singleton.get<RawConverter>()
                parameter = Singleton.get<ShibaFunctionExecutor>().execute(function, null)
            }
        }
        if (extensions.size == 1) {
            val extension = extensions.first()
            return when (val extensionValue = visit(extension, context)) {
                is ShibaBinding -> {
                    extensionValue.apply {
                        parameter = if (converter != null) {
                            ShibaConverterParameter(converter, parameter, function)
                        } else {
                            function
                        }
                        converter = Singleton.get<SingleBindingFunctionConverter>()
                    }
                }
                else -> {
                    ShibaBinding("").apply {
                        converter = Singleton.get<RawConverter>()
                        parameter = Singleton.get<SingleBindingShibaFunctionExecutor>().execute(function, extensionValue)
                    }
                }
            }
        }
        if (extensions.size > 1) {
            throw IllegalArgumentException("Currently only support single ShibaExtension")
        }
        throw IllegalArgumentException()
    }

    private fun getExtensions(function: ShibaFunction): List<ShibaExtension> {
        return function.parameter.map {
            when (it) {
                is ShibaExtension -> {
                    val executor = Shiba.configuration.extensionExecutors.firstOrNull { e -> e.name == it.type }
                    if (executor !is IMutableExtensionExecutor) {
                        arrayListOf(it)
                    } else {
                        null
                    }
                }
                is ShibaFunction -> getExtensions(it)
                else -> null
            }
        }.filter { it != null }.flatMap { it!! }
    }

    private fun parseFunction(function: ShibaFunction, context: IShibaContext?) : ShibaFunction {
        return function.apply {
            parameter = parameter.map {
                when (it) {
                    is ShibaFunction -> parseFunction(it, context)
                    is ShibaExtension -> {
                        val executor = Shiba.configuration.extensionExecutors.firstOrNull { e -> e.name == it.type }
                        if (executor is IMutableExtensionExecutor) {
                            executor.provideValue(context, it)
                        } else {
                            it
                        }
                    }
                    else -> {
                        val result = visit(it, context)
                        result
                    }
                }
            }.filter { it != null }.map { it!! }
        }
    }

    private fun visit(item: ObjectNode, context: IShibaContext?): ShibaObject {
        val map = ShibaObject()
        item.fieldNames().forEach {
            map[it] = visit(item.get(it), context)
        }
        return map
    }

    private fun visit(item: ArrayNode, context: IShibaContext?): List<Any?> {
        return item.map { visit(it, context) }
    }

    private fun visit(item: ValueNode, context: IShibaContext?): Any? {
        return item.run {
            when {
                isTextual -> textValue()
                isFloat -> floatValue()
                isDouble -> doubleValue()
                isInt -> intValue()
                isNumber -> numberValue()
                isBoolean -> booleanValue()
                isNull -> null
                else -> null
            }
        }
    }

    private fun visit(item: JSValue, context: IShibaContext?): Any? {
        return when {
            item.isObject -> return visitJSObject(item)
            item.isArray -> return visitJSArray(item)
            item.isBoolean -> item.toBoolean()
            item.isNumber -> item.toNumber()
            item.isString -> item.toString()
            item.isNull -> null
            else -> item
        }
    }

    private fun visitJSArray(item: JSValue): List<Any?> {
        return item.toJSArray().map { visit(it, null) }
    }

    private fun visitJSObject(item: JSValue): ShibaObject {
        val jsobj = item.toObject()
        val obj = ShibaObject()
        item.toObject().propertyNames().forEach {
            obj[it] = visit(jsobj.property(it), null)
        }
        return obj
    }
}

File: UWP/Shiba/Visitors/ValueVisitor.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using System.Reflection;
using System.Security.Cryptography;
using System.Text;
using Windows.UI.Xaml;
using ChakraHosting;
using Newtonsoft.Json.Linq;
using Shiba.Controls;
using Shiba.ExtensionExecutors;
using Shiba.Internal;
using Shiba.Scripting;
using Shiba.ViewMappers;
using ShibaView = Shiba.Controls.View;
using NativeView = Windows.UI.Xaml.FrameworkElement;
using NativeBinding = Windows.UI.Xaml.Data.Binding;
using NativeProperty = Windows.UI.Xaml.DependencyProperty;
using NativeThickness = Windows.UI.Xaml.Thickness;
using NativeViewGroup = Windows.UI.Xaml.Controls.Panel;

namespace Shiba.Visitors
{
    internal class ValueVisitor
    {
        public object DynamicVisit(object obj, IShibaContext context)
        {
            if (obj == null)
            {
                return null;
            }
            return Visit((dynamic) obj, context);
        }

        private object Visit(object obj, IShibaContext context)
        {
            return obj;
        }

        //TODO: Dictionary means it does not support multiple renderer for one view (like Xamarin.Forms custom renderer)
        private static readonly ConcurrentDictionary<string, IViewMapper> Renderer =
            new ConcurrentDictionary<string, IViewMapper>();
        private static readonly IViewMapper ShibaHostMapper = new ComponentMapper();

        private NativeView Visit(View view, IShibaContext context)
        {
            var attribute =
                ShibaApp.Instance.ViewMappings.LastOrDefault(it =>
                    view.ViewName == it.ViewName);
            if (attribute == null)
            {
                if (ShibaApp.Instance.Components.ContainsKey(view.ViewName))
                {
                    view.Properties.Add(new Property("componentName", view.ViewName));
                    return ShibaHostMapper.Map(view, context) as ShibaHost;
                }

                return null;
            }

            var renderer = Renderer.GetOrAdd(attribute.ViewName, type => CreateRenderer(attribute));
            if (context == null) throw new ArgumentNullException($"{nameof(context)} can not be null");

            var target = renderer.Map(view, context) as NativeView;
            if (view.Children.Any() && target is NativeViewGroup panel)
            {
                var commonprops = new List<(ShibaView view, NativeView native, List<Property> properties)>();
                view.Children.ForEach(it =>
                {
                    var child = Visit(it, context);
                    if (renderer is IAllowChildViewMapper viewGroupMapper)
                    {
                        viewGroupMapper.AddChild(panel, child);
                    }
                    else
                    {
                        panel.Children.Add(child);
                    }
                    var commonprop = it.Properties.Where(prop =>
                            ShibaApp.Instance.Configuration.CommonProperties.Any(cp => cp.Name == prop.Name))
                        .ToList();
                    if (commonprop.Any()) commonprops.Add((it, child, commonprop));
                });

                commonprops.ForEach(it =>
                {
                    var (view1, native, properties) = it;
                    properties.ForEach(prop =>
                    {
                        ShibaApp.Instance.Configuration.CommonProperties
                            .Where(cp => cp.Name == prop.Name).ToList()
                            .ForEach(cp => cp.Handle(prop.Value, view1, native, target));
                    });
                });
            }

            return target;
        }

        private static IViewMapper CreateRenderer(ExportMapperAttribute arg)
        {
            return Activator.CreateInstance(arg.MapperType) as IViewMapper;
        }
        
        private object Visit(ShibaExtension item, IShibaContext context)
        {
            var executor =
                ShibaApp.Instance.Configuration.ExtensionExecutors.FirstOrDefault(it => it.Name == item.Type);
            if (executor != null)
            {
                var value = executor.ProvideValue(context, item);
                if (string.IsNullOrEmpty(item.ScriptName)) return value;

                var func = new ShibaFunction(item.ScriptName) {Parameters = {item}};
                if (value is NativeBinding binding)
                {
                    if (binding.Converter != null)
                        binding.ConverterParameter = new ShibaConverterParameter
                        {
                            InnerConverter = binding.Converter,
                            InnerParameter = binding.ConverterParameter,
                            Function = func
                        };
                    else
                        binding.ConverterParameter = func;
                    binding.Converter = Singleton<SingleBindingFunctionConverter>.Instance;
                    return binding;
                }

                return new NativeBinding
                {
                    ConverterParameter =
                        Singleton<SingleBindingShibaFunctionExecutor>.Instance.Execute(func, item),
                    Converter = Singleton<RawDataConverter>.Instance
                };
            }

            throw new NotImplementedException();
        }
        
        private NativeBinding Visit(ShibaFunction item, IShibaContext context)
        {
            var function = ParseFunction(item, context);
            var extensions = GetExtensions(function)?.ToList();
            if (extensions == null || !extensions.Any())
                return new NativeBinding
                {
                    ConverterParameter = Singleton<ShibaFunctionExecutor>.Instance.Execute(function, null),
                    Converter = Singleton<RawDataConverter>.Instance
                };

            if (extensions.Count == 1)
            {
                var extension = extensions.FirstOrDefault();

                var extensionValue = DynamicVisit(extension, context);

                switch (extensionValue)
                {
                    case NativeBinding binding:
                        if (binding.Converter != null)
                            binding.ConverterParameter = new ShibaConverterParameter
                            {
                                InnerConverter = binding.Converter,
                                InnerParameter = binding.ConverterParameter,
                                Function = function
                            };
                        else
                            binding.ConverterParameter = function;
                        binding.Converter = Singleton<SingleBindingFunctionConverter>.Instance;
                        return binding;
                    default:
                        return new NativeBinding
                        {
                            ConverterParameter =
                                Singleton<SingleBindingShibaFunctionExecutor>.Instance.Execute(function,
                                    extensionValue),
                            Converter = Singleton<RawDataConverter>.Instance
                        };
                }
            }

            if (extensions.Count > 1)
                throw new ArgumentOutOfRangeException("Currently only support single ShibaExtension");

            throw new ArgumentOutOfRangeException($"Can not handle function call ${item}");
        }

        private IEnumerable<ShibaExtension> GetExtensions(ShibaFunction function)
        {
            foreach (var item in function.Parameters)
                switch (item)
                {
                    case ShibaExtension extension:
                    {
                        var executor = ShibaApp.Instance.Configuration.ExtensionExecutors
                            .FirstOrDefault(it => it.Name == extension.Type);
                        if (!(executor is IMutableExtensionExecutor)) yield return extension;
                    }
                        break;
                    case ShibaFunction shibaFunction:
                        foreach (var extension in GetExtensions(shibaFunction)) yield return extension;
                        break;
                }
        }

        private ShibaFunction ParseFunction(ShibaFunction item, IShibaContext context)
        {
            for (var i = 0; i < item.Parameters.Count; i++)
            {
                var parameter = item.Parameters[i];
                switch (parameter)
                {
                    case ShibaFunction function:
                        item.Parameters[i] = ParseFunction(function, context);
                        break;
                    case ShibaExtension extension:
                        var executor = ShibaApp.Instance.Configuration.ExtensionExecutors
                            .FirstOrDefault(it => it.Name == extension.Type);
                        if (executor is IMutableExtensionExecutor)
                            item.Parameters[i] = executor.ProvideValue(context, extension);
                        break;
                    default:
                        item.Parameters[i] = DynamicVisit(parameter, context);
                        break;
                }
            }

            return item;
        }


        private List<object> Visit(JArray obj, IShibaContext context)
        {
            return obj.Select(it => DynamicVisit(it, context)).ToList();
        }

        private ShibaObject Visit(JObject item, IShibaContext context)
        {
            var dic = new ShibaObject();
            foreach (var (key, value) in item)
            {
                dic.TryAdd(key, DynamicVisit(value, context));
            }

            return dic;
        }

        private KeyValuePair<string, object> Visit(JProperty item, IShibaContext context)
        {
            return new KeyValuePair<string, object>(item.Name, DynamicVisit(item.Value, context));
        }

        private object Visit(JValue item, IShibaContext context)
        {
            switch (item.Type)
            {
                case JTokenType.Integer:
                    return item.Value<int>();
                case JTokenType.Float:
                    return item.Value<float>();
                case JTokenType.String:
                    return item.Value<string>();
                case JTokenType.Boolean:
                    return item.Value<bool>();
                case JTokenType.Date:
                    return item.Value<DateTime>();
                case JTokenType.Raw:
                    return item.Value<string>();
                case JTokenType.Bytes:
                    return item.Value<byte[]>();
                case JTokenType.Guid:
                    return item.Value<Guid>();
                case JTokenType.Uri:
                    return item.Value<string>();
                case JTokenType.TimeSpan:
                    return item.Value<TimeSpan>();
                default:
                    return null;
            }
        }

        private object Visit(JavaScriptValue value, IShibaContext context)
        {
            switch (value.ValueType)
            {
                case JavaScriptValueType.Error:
                case JavaScriptValueType.Object:
                    return VisitJavascriptObject(value);
                case JavaScriptValueType.Array:
                    return VisitJavascriptArray(value);
                case JavaScriptValueType.Boolean:
                    return value.ToBoolean();
                case JavaScriptValueType.Null:
                    return null;
                case JavaScriptValueType.Number:
                    return value.ToDouble();
                case JavaScriptValueType.String:
                    return value.ToString();
                case JavaScriptValueType.Undefined:
                case JavaScriptValueType.Function:
                default:
                    return value;
            }
        }

        private List<object> VisitJavascriptArray(JavaScriptValue value)
        {
            var result = new List<object>();
            {
                var currentIndex = 0;
                while (value.HasIndexedProperty(currentIndex.ToJavaScriptValue()))
                {
                    result.Add(Visit(value.GetIndexedProperty(currentIndex.ToJavaScriptValue()), null));
                    currentIndex++;
                }
            }
            return result;
        }

        private ShibaObject VisitJavascriptObject(JavaScriptValue value)
        {
            var propers = VisitJavascriptArray(value.GetOwnPropertyNames()).Cast<string>().ToList();
            var obj = new ShibaObject();
            foreach (var name in propers)
            {
                obj.TryAdd(name, Visit(value.GetProperty(name.ToJavaScriptPropertyId()), null));
            }

            return obj;
        }
    }

}